sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala

```diff
@@ -248,6 +275,71 @@ case class OptimizeSkewedJoin(conf: SQLConf) extends Rule[SparkPlan] {
       } else {
         smj
       }
+
+    case bnl @ BroadcastNestedLoopJoinExec(
+    ShuffleStage(left: ShuffleStageInfo), _, BuildRight, joinType, _, _) =>
+      resolveBroadcastNLJoinSkew(bnl, left, joinType, BuildRight)
+    case bnl @ BroadcastNestedLoopJoinExec(
+    _, ShuffleStage(right: ShuffleStageInfo), BuildLeft, joinType, _, _) =>
+      resolveBroadcastNLJoinSkew(bnl, right, joinType, BuildLeft)
+  }
+
+  def resolveBroadcastNLJoinSkew(
+      bnl: BroadcastNestedLoopJoinExec,
+      stream: ShuffleStageInfo,
+      joinType: JoinType,
+      buildSide: BuildSide): SparkPlan = {
+    val streamMedSize = medianSize(stream.mapStats)
+    val numPartitions = stream.partitionsWithSizes.length
+    logDebug(
+      s"""
+         |Optimizing skewed join.
+         |${if (buildSide == BuildLeft) "Left" else "Right"} stream side partitions size info:
+         |${getSizeInfo(streamMedSize, stream.mapStats.bytesByPartitionId)}
+        """.stripMargin)
+    val streamActualSizes = stream.partitionsWithSizes.map(_._2)
+    val streamTargetSize = targetSize(streamActualSizes, streamMedSize)
+    val streamSidePartitions = mutable.ArrayBuffer.empty[ShufflePartitionSpec]
+    var numSkewedStream = 0
+    for (partitionIndex <- 0 until numPartitions) {
```

Comment:
This part looks like that in `optimizeSkewJoin`. Can we make it into a general method ?